Handle root being deleted by setting the root to the parent directory.

bf319 opened the pull request:
If the root is deleted, set its parent as the new root.

When a file is deleted, select its parent in the file tree so that keyboard navigation in the file tree remains active.

irengrig commented on src/vs/workbench/contrib/files/browser/fileActions.ts:
```diff
@@ -972,7 +972,20 @@ export const deleteFileHandler = async (accessor: ServicesAccessor) => {
 
 	if (stats.length) {
-		await deleteFiles(accessor.get(IWorkingCopyFileService), accessor.get(IDialogService), accessor.get(IConfigurationService), stats, false);
+		const lastItemDeleted = stats[stats.length - 1];
```
Is the list of selected items ordered?
What I with multiple selection I select the subdirectory and all its children - in that case, will the last deleted be the last child of the subdirectory? It seems the subdirectory itself should be put into lastItemDeleted for transferring selection...

bf319 replied:
I did not realize that the deletion can be handled in explorerService. I definitely it would be better to do it there.

I need to look a little more into explorerService, but from what I've seen so far the list is ordered and if a directory and all its children are deleted, the handler is only triggered for the directory.

irengrig commented on src/vs/workbench/contrib/files/browser/fileActions.ts:
```diff
@@ -972,7 +972,20 @@ export const deleteFileHandler = async (accessor: ServicesAccessor) => {
+		deleteFiles(accessor.get(IWorkingCopyFileService), accessor.get(IDialogService), accessor.get(IConfigurationService), stats, false).then(() => {
+			// If the user deletes the focused directory, focus on the immediate parent directory
+			if (stats.includes(currentRoot) && !explorerService.findClosest(currentRoot.resource)) {
+				explorerService.setRoot(resources.dirname(currentRoot.resource));
```
Will dirname() for the directory return the directory itself? (I am judging by the unix command)
We should somehow get the Uri of the parent - maybe with URI.fsPath() get the file system path and operate that.

bf319 replied:
dirname(resource) returns the URI of the directory containing 'resource'. This works regardless of whether resource points to a file or a directory.

I think there are multiple ways to get to the parent resource, but from what I've seen so far in the code, dirname() is often used (e.g. in onDidRunOperation() in explorerService).

bf319 commented:
In explorerModel:
   Added sortOrder (passed from explorerService) to call to setRoot() so that the order of the children seen in the model is preserved.

In files.ts:
   Added findAdjacentSibling() method to IExplorerView interface so that it can be used in explorerService.

In explorerView.ts:
   Implemented findAdjacentSibling() which, given a stat: ExplorerItem, returns the sibling right above it in the explorer, or if none exists (element is the first child of the directory), the sibling at position 1. If this does not exist either, the parent directory will have no children when this element is deleted, so move the focus to the parent.

   Changed call in onDidActiveEditorChange() listener so that if the active file gets deleted, the sibling of this file is selected rather than having no selection at call.

In explorerService.ts:
   Handled deletion of current root by setting the root to the parent of this directory.
   Handled deletion of a file by selecting its sibling in the explorer so keyboard navigation can continue.

irengrig commented on src/vs/workbench/contrib/scopeTree/browser/explorerView.ts:
```diff
@@ -673,6 +673,27 @@ export class ExplorerView extends ViewPane {
+		const orderedChildren: ExplorerItem[] = [];
+		const parent = deleted.parent;
+
+		this.tree.getNode(parent).children.forEach(e => {
```
It is possible to use just a single iteration here without storing orderedChildren.

bf319 replied:
Got it, thank you! I've pushed a change for this.
